### PR TITLE
Single-Ayah Viewing

### PR DESCRIPTION
**Pull request by muhsinali**
Builds clean. Removed pagers and it looks nice.

We should think about if we're going to remove / condense the header or change styles, however this PR should serve as a functional update. Still supports custom queries like "2/2-21"

### CODE REVIEW

File: src/scripts/components/header/MasterHeader.js [hunk 1]
```diff
@@ -23,10 +23,9 @@ class MasterHeader extends React.Component{
     this.setState({showOptions: !this.state.showOptions});
   };
 
-  // Disable hyperlink for previous surah if surah is Al-Faatihah.
   previousChapter() {
-    var prev = '/' + (parseInt(this.props.currentRoute.get('params').get('surahId')) - 1);
     var currentPos = parseInt(this.props.currentRoute.get('params').get('surahId'));
+    var prev = '/' + (currentPos - 1);
```
Comment (mmahalwy) [inline src/scripts/components/header/MasterHeader.js]: Good catch :)

File: src/scripts/configs/routes.js [hunk 2]
```diff
@@ -71,14 +71,15 @@ export default {
...
         () => {
-          if (currentRoute.get('params').get('range')) {
-            if (currentRoute.get('params').get('range').indexOf('-') > -1) {
-              fromParam = currentRoute.get('params').get('range').split('-')[0];
-              toParam = currentRoute.get('params').get('range').split('-')[1];
-            }
-            else {
-              fromParam = currentRoute.get('params').get('range');
-              toParam = parseInt(fromParam) + 10
+          var range = currentRoute.get('params').get('range');
```
Comment (mmahalwy) [inline src/scripts/configs/routes.js]: Nice :)

**Comment by mmahalwy** [comment]
We probably have to tell lazy loading to stop loading... you're probably getting errors in the console or something weird is happening. See: https://github.com/quran/quran.com-frontend/blob/master/src/scripts/routes/Surah.js#L88

**Comment by mmahalwy** [comment]
Also, I think this will error out the audioplayer or make it go crazy!

**Comment by muhsinali** [comment]
>We probably have to tell lazy loading to stop loading... you're probably getting errors in the console or >something weird is happening.

I'm currently not getting any errors or feedback when scrolling in a single ayah so I will definitely check it out to make sure that kind of background process isn't running. 

>Also, I think this will error out the audioplayer or make it go crazy! 

You're absolutely right the next step IMO would be to think about what should be done with the header / audioplayer. It seems that entire div is showing like normal but not functioning. In the PT there was a note about possibly not showing audio or something of that nature so I wasn't sure which way to go

**Comment by mmahalwy** [comment]
Yeah i figured the audioplayer would freak out. Hmmm.. I am not sure, we should has Umar, the designer.

**Comment by muhsinali** [comment]
When debugging I'm not getting errors while scrolling in neither my npm console or JS debug, did you test this locally?

**Comment by mmahalwy** [comment]
Nope ill test soon!

**Comment by muhsinali** [comment]
Any ideas about what we're going to do with this?

**Comment by mmahalwy** [comment]
@muhsinali not sure yet! LEt's make this happen :)

**Comment by muhsinali** [comment]
@mmahalwy Hey can you check this out? This is running on my end with no errors. This needs to be tested because even on master the audio is only playing 1 ayah, and that's without any of these changes. Before the entire audio was breaking when reading single ayah however now it's playing cleanly with no errors being logged that I could see. Let me know!!!

**Comment by muhsinali** [comment]
Updated to be able to link to a single surah in a search

link

**Comment by mmahalwy** [comment]
Ill check this out shortly! Dw :)

**Comment by mmahalwy** [comment]
Looks and feels great mashallah! 
Love it! 

Only thing: once the audio is finished, it results to this:
image

Means that it wants the next ayah and cannot find it because I load the next ayahs when we are 3 ayahs away from the last ayah. One way to resolve this is [here](https://github.com/quran/quran.com-frontend/blob/master/src/scripts/components/audioplayer/Audioplayer.js#L112) just add a check to see if only one ayah loaded, if yes, then don't change to next ayah. If no, change to next ayah. 

;)

**Comment by muhsinali** [comment]
I was actually waiting for this exact thing :D I think my problem with audio is probably fixed with the patch for Firefox & Safari ... I'm on Firefox so I was getting nothing at all after the first ayah, not even a console msg. I'll get this fixed today insha'Allah.

On a side note has anyone tested lately on IE? With a lot of changes happening and IE having a history of not working it can be worrisome

**Comment by muhsinali** [comment]
Would this work? When trying to merge w/ the Firefox patch it completely freaked out because there have been a lot of changes.

**Comment by muhsinali** [comment]
I've posted the change and currently working through the merge conflicts ... I hate merge conflicts :D

**Comment by mmahalwy** [comment]
just do `git pull --rebase origin master` and work through them :)
Merging is a devil lol

**Comment by muhsinali** [comment]
Code updated! Audio player is working and shuts down when listening to single ayah and resets the play button.

**Comment by muhsinali** [comment]
When doing custom ayah views, so if the user types "quran.com/2/4-8" it's still coming back as currentAyah is undefined. I'm not sure how to test for this scenario. It is in fact working and stopping at the correct spots but it is indeed logging the error. Is there a way we can test for the entered URL from the get request and just play THAT? Ideas?

**Comment by mmahalwy** [comment]
@muhsinali interesting find. I think this could be due to the number of ayahs it paginates afterwards. What happens when you do /2/1-10 and get to the last ayah? Still error?

File: src/scripts/stores/AyahsStore.js [hunk 3]
```diff
@@ -153,7 +153,13 @@ class AyahsStore extends BaseStore {
     this.searchStats = state.searchStats;
 
     if (!!~~state.ayahs.length) {
-      this.buildAudio([this.ayahs[0], this.ayahs[1]]);
+      // Fixed the audio breaking if a single ayah was passed because
```
Comment (mmahalwy) [inline src/scripts/stores/AyahsStore.js]: Nice :)

**Comment by mmahalwy** [comment]
Hmm: https://github.com/quran/quran.com-frontend/pull/12/files#diff-69f7667d76e50024f3c34d5f4966beb5R184

**Comment by muhsinali** [comment]
> ...due to the number of ayahs it paginates afterwards. 

I agree because at the moment it's only checking for *one*, then stopping after that accordingly because we're explicitly testing for it. The problem is, is how to test for the exact amount the user put in, so if they want to view 1, 4, or 56 ayaat at a time the player needs to stop ... right?

**Comment by muhsinali** [comment]
@mmahalwy There may be a waaaay simpler resolution than this but would it make sense to receive the get request info, and if it doesn't have an ayah (range) ex. 'quran.com/5' conduct the normal logic. If it does have a range 'quran.com/5/6-15' to load the first three, like normal, and if the currentAyah is equal to '15', shut down the player? 

Just brainstorming here there's probably something much simpler :D I tend to overthink alhamdulillah.

**Comment by mmahalwy** [comment]
@muhsinali no no, the audioplayer has to know when to load the next ayahs, giving the user a seamless experience. So when I am at 5/6-15 and I reach 15, it should play 16... but to get 16, we need to know the range between 6 and 15, and then load them before the user gets to 16. i do this at 3 ayahs before the end (so 12). 

With single view, we need to stop the 'load more' function which is in the AudioplayerActions.

See
image

The actions are responsible (or at least I try to put that logic there) for determining what to do. So AyahsActions and AudioplayerActions, etc. determine what to do given I have a certain set of information, and the store is good at storing them. They may need to do some logic but mostly to store things. The components are suppose to be stupid and not think things. It should get information from the store and go from there.

I do apologize about this, I am not trying to give you a hard time and I hope you don't ever think so - I actually want you to struggle with this feature a little bit to understand Reactjs and Flux better - it's only going to make you much better. And it's also a great opportunity to ask lots of questions and for me to answer them to help your understanding :)

**Comment by muhsinali** [comment]
@mmahalwy Not a problem, I understand. I'll definitely have to think of some logic here because we do run into a small quandary. However my question is do you think the user would expect the surah & audio player to load more if they hard type "3/4-17" after finishing 17? We could still load 3 when we're 3 ayaat in advance unless we're getting to the end, in which it would not load more and would stop after playing 17. This would still be the same seamless operation, just checking in advance when to stop loading. If the user just types "quran.com/3" this shouldn't be affected at all and would load on as much as the reader would like, we would only be checking for set ayaat values passed in.

I personally don't know another way to do it because we would run into a loop of sorts. We want to keep loading seamlessly however if the user only wants a set amount of ayaat the player needs to stop there...

**Comment by mmahalwy** [comment]
I think the user would want to continue listening, for example, I listened to half of surah 2 yesterday and want to continue listening today. Or I was reading or learning and now want to continue. BOOM! 

I think the audioplayeractions just needs to check if there is 1 ayah in the ayahsstore or more, if 1, don't load more. 
:)

**Comment by muhsinali** [comment]
@mmahalwy I can't seem to figure out the best way here. With the code as it is here I'm checking if only 1 ayah is being fed in. The problem I'm running into is whenever a uses passes a different set of custom ayaat it plays like normal, however logs an error message after playing the last ayah. 

I think we're on 2 different wavelengths in terms of what's happening though. I don't know how to check to see if the player should stop executing but at the same time not look at what the user passed in as values. Wouldn't checking to see if there is 1 ayah in the ayahstore do the same thing the code is doing now? If there is 1 ayah passed in '5/67' it indeed plays that and only that ayah with no errors as it is. The problem is who knows how many ayaat the user wants to see? Yani I'm testing for one ayah, and stopping the player after it finishes. However if they pass in two it's running as normal and trying to fetch more ayaat. Without looking at the passed in values I would need help figuring this out. 

If you could help I would appreciate it I'm kind of lost here ... sorry if I'm being a pain akhi :\

**Comment by muhsinali** [comment]
@mmahalwy Any further ideas on this? It's currently working with a single ayah but I need your expertise ya akhi

**Comment by mmahalwy** [comment]
@muhsinali going to check into this tonight and lets get it merged then

**Comment by muhsinali** [comment]
@mmahalwy Sounds good, just to be clear do you want me to merge it now or after you check it out?

**Comment by mmahalwy** [comment]
@muhsinali any update on this?

**Comment by muhsinali** [comment]
@mmahalwy No, last thing posted was that you were going to check it out. I've had to take a break until recently and now there's merge conflicts so I don't know what's going on really.

**Comment by mmahalwy** [comment]
Okay no worries. I'll hop on this to get it merge fixed

**Comment by mmahalwy** [comment]
Ill send it your way for review inshallah

**Comment by muhsinali** [comment]
Insha'allah. I've been meaning to get in touch with you to see how things
were going I was wanting to get back to helping again things got a little
hectic here.

**Comment by mmahalwy** [comment]
@muhsinali no worries! I understand :) Let me know when you're ready to come back

**Comment by mmahalwy** [comment]
Closing this for favor of new branch

**Pull request closed without merging**